addon/utils/query-cache.js

```diff
@@ -0,0 +1,69 @@
+
+const CACHE_DECAY_TIMEOUT = 5000;
+
+export default class QueryCache {
+  constructor() {
+    this._internalCache = Object.create(null);
+    this.decayId = null;
+  }
+
+  add(type, key, value) {
+    if (!key || !value) {
+      return;
+    }
+
+    if (!this._internalCache[type]) {
+      this._internalCache[type] = Object.create(null);
+    }
+
+    this._internalCache[type][key] = value;
+  }
+
+  remove(type, key) {
+    let typeMap = this._internalCache[type];
+
+    if (typeMap && typeMap[key]) {
+      delete typeMap[key];
+    }
+  }
+
+  clear() {
+    this._internalCache = Object.create(null);
+  }
+
+  get(type, key) {
+    if (!this._internalCache[type]) {
+      return;
+    }
+
+    return this._internalCache[type][key];
+  }
+
+  get types() {
+    return Object.keys(this._internalCache);
+  }
+
+  scheduleDecay(delay = CACHE_DECAY_TIMEOUT, onDecay) {
```

Comment:
Fair :)

We plan on restricting number of entries in the cache + it's scoped to the route. I see it could get complicated

The example of 3 api calls are a use case we need today :)


The example of the